### ISSUE TITLE
Extract Com Params from smr-d

### ISSUE THREAD

**Issue by Flo354** [jglim/ODB#9]
Hello,

I studied the program because I would like to automate the extraction of com params for a personal project. For instance, `CP_UniqueRespIdTable`, `CP_StMin`...

I noticed in the strings extracted that there are these parameters, but they are not always accurate. Let's take for example two ECUs for the UniqueRespIdTable:
- MED1775 => [MED1775.smr-d-string.txt](https://github.com/user-attachments/files/16783773/MED1775.smr-d-string.txt)
- BC_F222 => [BC_F222.smr-d-string.txt](https://github.com/user-attachments/files/16783771/BC_F222.smr-d-string.txt)

In the case of MED1775, you will see something like this:

```
ISO_15765_2.smr-cs::ISO_15765_2.CP_UniqueRespIdTable
normal segmented 11-bit transmit with FC
2016
normal segmented 11-bit receive with FC
2024
normal unsegmented 11-bit receive
UDS_CAN_D
```

As expected, we can see the can ID and the tester ID

But for BC_F222, this is a complete different story:

```
ISO_15765_2.smr-cs::ISO_15765_2.CP_UniqueRespIdTable
normal segmented 11-bit transmit with FC
normal segmented 11-bit receive with FC
1417
normal unsegmented 11-bit receive
UDS_CAN_D
```

Here, we have only the tester ID. The can ID should be below `normal segmented 11-bit transmit with FC` but it is not.
This is one example. On some ECUs, the UniqueRespIdTable is not present, or both of the can ID and tester ID are not present.

For BC_F222, we can still find the can ID. going through database differ we can see that the can id is "1545"
![image](https://github.com/user-attachments/assets/299fb00a-db3a-4295-938c-f0c96b82bce1)

The line \00 1545 \00 (or \n if replaced) is well defined:
![image](https://github.com/user-attachments/assets/c132f098-b300-4b7e-8f98-cd3a1be20ece)

But I am yet to understand how can I parse this accordingly.

Do you have any idea how can I parse this accurately as Database Differ does please?

Thanks,

**Comment by jglim** [comment]
Hello Florian,

I did not look further into the format after I decrypted the main sections. Within your SMR files, there are..

- unknown first block
- binary block, where DLLs, JARs can be carved from
- strings block
- optional flash block, specific to smr-f

https://github.com/jglim/ODB/blob/783e666dd612106b1abf8566c390fc482ed729dc/ODB/ObjectDB/ODBFile.cs#L172

Your attachments come from the strings block, which seems to purely consist of a string table. I assume that one of the earlier 2 blocks contains a tree-like structure which is what we see in DatabaseDiffer. For string fields in that structure, there is probably a pointer or offset to the string table.

When the strings are serialized, I am fairly certain that there is deduplication involved (esp. since many variants share identical values), which should explain why some values are missing. 

Unfortunately beyond this, I do not know more about the tree structure. I am very interested in knowing how it works too, but I am also stuck at this point.

**Comment by Flo354** [comment]
> When the strings are serialized, I am fairly certain that there is deduplication involved (esp. since many variants share identical values), which should explain why some values are missing.

This is actually very interesting and can explain what I am facing.

Do you have any clue on how we could eventually be able to tackle this? How did you find about the decryption process? Reverse engineering? Resources on the web? ...

**Comment by jglim** [comment]
This project was implemented from reversing DTS, specifically version 8.14. The bulk of the interesting bits (such as decryption, keys, hashes, segments) were found in `ODBase.dll`. I am not sure if the higher-level parsing is still contained within this DLL.

I don't have any good ideas on moving forward, but DatabaseDiffer (and the DLLs that it depends on) is a good candidate to dig into, since DatabaseDiffer is relatively lightweight as compared to Monaco.

**Comment by Flo354** [comment]
I found something:

`ODBase.dll` is embedding `StdUtil.dll` which itself is including these three DLLs:
- AvlTrees.dll => Proprietary DLL for parsing AVL Tree => https://en.wikipedia.org/wiki/AVL_tree
- Xalan-C_1_11.dll => Transform XML documents using XSLT => https://xalan.apache.org/index.html
- xerces-c_3_1.dll => XML Parser => https://xerces.apache.org/index.html

First guess: the unknown data section is an AVL Tree which is used to regenerate the XML ODB file.

**Comment by jglim** [comment]
That's interesting. I'm struggling to understand if it is possible and practical to fit an XML file into an AVL tree. The XSLT transformation does make sense though, but that probably comes after the first layer is removed?

TrafoCoApp has also been a mystery to me. I haven't been able to get it do achieve anything meaningful, and there's little documentation for that application.

**Comment by Flo354** [comment]
I think the AVL is somehow a mapper tree to the string tree.
I guess there must be a path in the binary tree which when you follow it guides you toward the correct string.
Since it's binary search, that would explain why the needed strings are not always at the same place.

We need now to understand how to parse this AVL tree and how to map it in order to rebuild the XML ODX.

I think this is where TrafoCoApp.exe is useful. In my opinion, this app is used somehow to generate smr files from the OTX Studio. Maybe it would be easier trying to convert ODX to SMR. Since it must be almost the same operation, when we know how to go from A to B, it would be easier to go from B to A.

What do you think?

**Comment by jglim** [comment]
If I had to guess, strings in the tree are a bunch of offsets pointing at the strings blob. Non-primitive types (e.g. byte arrays) are probably embedded in the other binary blob in the same manner. Primitive fixed-length types e.g. int32 could possibly be embedded directly in the tree. I do not know, or have any assumptions about the actual tree format, though there's a good chance it is embedded in the first unknown block.

However, this is mostly guesswork. I got stuck here on static RE; running DTS applications with an attached debugger might yield more useful leads.

**Comment by Flo354** [comment]
Yes, this is out of my field either, I will continue to extract data through Database Differ :p
